Refactor autocast C++ APIs to be device-agnostic

guangyey opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124479
* __->__ #124359

# Motivation
This PR aims to refactor autocast **C++** APIs to be device-agnostic and deprecate the device-specific autocast  **C++** APIs.
In C++ side,
- `is_enabled()` -> `is_enabled(device_type)`.
- `set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
- `get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
- `set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`

These following C++ APIs are deprecated and should be removed in PyTorch 2.5
- `is_cpu_enabled`
- `set_cpu_enabled`
- `get_autocast_cpu_dtype`
- `set_autocast_cpu_dtype`
- `is_xpu_enabled`
- `set_xpu_enabled`
- `get_autocast_xpu_dtype`
- `set_autocast_xpu_dtype`
- `is_ipu_enabled`
- `set_ipu_enabled`
- `get_autocast_ipu_dtype`
- `set_autocast_ipu_dtype`
- `is_hpu_enabled`
- `set_hpu_enabled`
- `get_autocast_hpu_dtype`
- `set_autocast_hpu_dtype`
- `is_xla_enabled`
- `set_xla_enabled`
- `get_autocast_xla_dtype`
- `set_autocast_xla_dtype`
- `is_privateuseone_enabled`
- `set_privateuseone_enabled`
- `get_autocast_privateuseone_dtype`
- `set_autocast_privateuseone_dtype`

In Python side,
provide 4 generic autocast APIs:
- `torch.is_autocast_enabled(device_type)`
- `torch.set_autocast_enabled(device_type, new_enabled)`
- `torch.get_autocast_dtype(device_type)`
- `torch.set_autocast_dtype(device_type, dtype)`

# Additional Context
We will submit another PR to refactor autocast **Python** APIs based on this PR.

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

guangyey commented:
@albanD This PR is working in progress. Is this PR reasonable because it is BC-breaking? If it is reasonable, I will continue to finish it. Could you give me your input?

guangyey committed:
refactor autocast APIs to be device-agnostic

[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Refactor autocast APIs to be device-agnostic"




cc mcarilli ptrblck leslie-fang-intel jgong5

[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Refactor autocast APIs to be device-agnostic"




cc mcarilli ptrblck leslie-fang-intel jgong5

[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Refactor autocast APIs to be device-agnostic"




cc mcarilli ptrblck leslie-fang-intel jgong5

[ghstack-poisoned]

albanD commented:
I definitely love the direction!
We do want to be a bit careful with public API changes though:
- In c++ if a github search doesn't return any hit, then I'm happy with making the BC-breaking change in this PR.
- In python, we should deprecate all APIs for at least one version (if mostly un-used) or two version (if we see significant use via github search). The full process would be:
  - This PR can add the new API
  - This PR can add a warning about deprecation for the old API
  - For APIs only needing one version, just after the 2.4 branch cut, we can do a follow up deleting the deprecated APIs
  - For APIs needing two versions, wait after the 2.5 branch cut for the follow up.

guangyey committed:
Update on "[WIP] Refactor autocast APIs to be device-agnostic"




cc mcarilli ptrblck leslie-fang-intel jgong5

[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Refactor autocast APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast APIs to be device-agnostic and remove the device-specific autocast APIs.
In C++ side,
`is_enabled()` -> `is_enabled(device_type)`.
`set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
`get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
`set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`
~~`is_cpu_enabled`~~
~~`set_cpu_enabled`~~
~~`get_autocast_cpu_dtype`~~
~~`set_autocast_cpu_dtype`~~
~~`is_xpu_enabled`~~
~~`set_xpu_enabled`~~
~~`get_autocast_xpu_dtype`~~
~~`set_autocast_xpu_dtype`~~
~~`is_ipu_enabled`~~
~~`set_ipu_enabled`~~
~~`get_autocast_ipu_dtype`~~
~~`set_autocast_ipu_dtype`~~
~~`is_hpu_enabled`~~
~~`set_hpu_enabled`~~
~~`get_autocast_hpu_dtype`~~
~~`set_autocast_hpu_dtype`~~
~~`is_xla_enabled`~~
~~`set_xla_enabled`~~
~~`get_autocast_xla_dtype`~~
~~`set_autocast_xla_dtype`~~
~~`is_privateuseone_enabled`~~
~~`set_privateuseone_enabled`~~
~~`get_autocast_privateuseone_dtype`~~
~~`set_autocast_privateuseone_dtype`~~

In Python side,
provide 4 generic autocast APIs:
-`torch.is_autocast_enabled(device_type)`
-`torch.set_autocast_enabled(device_type, new_enabled)`
-`torch.get_autocast_dtype(device_type)`
-`torch.set_autocast_dtype(device_type, dtype)`


cc mcarilli ptrblck leslie-fang-intel jgong5

[ghstack-poisoned]

guangyey commented:
> I definitely love the direction! We do want to be a bit careful with public API changes though:
> 
> * In c++ if a github search doesn't return any hit, then I'm happy with making the BC-breaking change in this PR.
> * In python, we should deprecate all APIs for at least one version (if mostly un-used) or two version (if we see significant use via github search). The full process would be:
>   
>   * This PR can add the new API
>   * This PR can add a warning about deprecation for the old API
>   * For APIs only needing one version, just after the 2.4 branch cut, we can do a follow up deleting the deprecated APIs
>   * For APIs needing two versions, wait after the 2.5 branch cut for the follow up.

Thanks for your elaboration. I capture your point.

guangyey committed:
Update on "[WIP] Refactor autocast APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast APIs to be device-agnostic and remove the device-specific autocast APIs.
In C++ side,
`is_enabled()` -> `is_enabled(device_type)`.
`set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
`get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
`set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`
~~`is_cpu_enabled`~~
~~`set_cpu_enabled`~~
~~`get_autocast_cpu_dtype`~~
~~`set_autocast_cpu_dtype`~~
~~`is_xpu_enabled`~~
~~`set_xpu_enabled`~~
~~`get_autocast_xpu_dtype`~~
~~`set_autocast_xpu_dtype`~~
~~`is_ipu_enabled`~~
~~`set_ipu_enabled`~~
~~`get_autocast_ipu_dtype`~~
~~`set_autocast_ipu_dtype`~~
~~`is_hpu_enabled`~~
~~`set_hpu_enabled`~~
~~`get_autocast_hpu_dtype`~~
~~`set_autocast_hpu_dtype`~~
~~`is_xla_enabled`~~
~~`set_xla_enabled`~~
~~`get_autocast_xla_dtype`~~
~~`set_autocast_xla_dtype`~~
~~`is_privateuseone_enabled`~~
~~`set_privateuseone_enabled`~~
~~`get_autocast_privateuseone_dtype`~~
~~`set_autocast_privateuseone_dtype`~~

In Python side,
provide 4 generic autocast APIs:
-`torch.is_autocast_enabled(device_type)`
-`torch.set_autocast_enabled(device_type, new_enabled)`
-`torch.get_autocast_dtype(device_type)`
-`torch.set_autocast_dtype(device_type, dtype)`


cc mcarilli ptrblck leslie-fang-intel jgong5

[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Refactor autocast APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast APIs to be device-agnostic and remove the device-specific autocast APIs.
In C++ side,
`is_enabled()` -> `is_enabled(device_type)`.
`set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
`get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
`set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`
~~`is_cpu_enabled`~~
~~`set_cpu_enabled`~~
~~`get_autocast_cpu_dtype`~~
~~`set_autocast_cpu_dtype`~~
~~`is_xpu_enabled`~~
~~`set_xpu_enabled`~~
~~`get_autocast_xpu_dtype`~~
~~`set_autocast_xpu_dtype`~~
~~`is_ipu_enabled`~~
~~`set_ipu_enabled`~~
~~`get_autocast_ipu_dtype`~~
~~`set_autocast_ipu_dtype`~~
~~`is_hpu_enabled`~~
~~`set_hpu_enabled`~~
~~`get_autocast_hpu_dtype`~~
~~`set_autocast_hpu_dtype`~~
~~`is_xla_enabled`~~
~~`set_xla_enabled`~~
~~`get_autocast_xla_dtype`~~
~~`set_autocast_xla_dtype`~~
~~`is_privateuseone_enabled`~~
~~`set_privateuseone_enabled`~~
~~`get_autocast_privateuseone_dtype`~~
~~`set_autocast_privateuseone_dtype`~~

In Python side,
provide 4 generic autocast APIs:
-`torch.is_autocast_enabled(device_type)`
-`torch.set_autocast_enabled(device_type, new_enabled)`
-`torch.get_autocast_dtype(device_type)`
-`torch.set_autocast_dtype(device_type, dtype)`


cc mcarilli ptrblck leslie-fang-intel jgong5

[ghstack-poisoned]

guangyey commented on torch/_C/__init__.pyi.in:
```diff
@@ -1269,8 +1269,16 @@ def is_grad_enabled() -> _bool: ...
 def _set_fwd_grad_enabled(enabled: _bool) -> None: ...
 def _is_fwd_grad_enabled() -> _bool: ...
 def is_inference_mode_enabled() -> _bool: ...
+@overload
```
Here add these overloads(new signature and deprecated signature) to pass CI, in the subsequent PR(refactor autocast python APIs), we will remove the deprecated signatures to prevent the developer from using them.

albanD replied:
Not sure I 100% follow this. but we should keep typing and docs for deprecated APIs (with appropriate warning in doc and when it runs) until we remove them!

guangyey replied:
OK, I will keep these deprecated signatures until they are removed in PyTorch 2.5.

guangyey committed:
Update on "[WIP] Refactor autocast APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast APIs to be device-agnostic and remove the device-specific autocast APIs.
In C++ side,
`is_enabled()` -> `is_enabled(device_type)`.
`set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
`get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
`set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`
~~`is_cpu_enabled`~~
~~`set_cpu_enabled`~~
~~`get_autocast_cpu_dtype`~~
~~`set_autocast_cpu_dtype`~~
~~`is_xpu_enabled`~~
~~`set_xpu_enabled`~~
~~`get_autocast_xpu_dtype`~~
~~`set_autocast_xpu_dtype`~~
~~`is_ipu_enabled`~~
~~`set_ipu_enabled`~~
~~`get_autocast_ipu_dtype`~~
~~`set_autocast_ipu_dtype`~~
~~`is_hpu_enabled`~~
~~`set_hpu_enabled`~~
~~`get_autocast_hpu_dtype`~~
~~`set_autocast_hpu_dtype`~~
~~`is_xla_enabled`~~
~~`set_xla_enabled`~~
~~`get_autocast_xla_dtype`~~
~~`set_autocast_xla_dtype`~~
~~`is_privateuseone_enabled`~~
~~`set_privateuseone_enabled`~~
~~`get_autocast_privateuseone_dtype`~~
~~`set_autocast_privateuseone_dtype`~~

In Python side,
provide 4 generic autocast APIs:
-`torch.is_autocast_enabled(device_type)`
-`torch.set_autocast_enabled(device_type, new_enabled)`
-`torch.get_autocast_dtype(device_type)`
-`torch.set_autocast_dtype(device_type, dtype)`


cc mcarilli ptrblck leslie-fang-intel jgong5

[ghstack-poisoned]

jgong5 commented on aten/src/ATen/autocast_mode.cpp:
```diff
@@ -91,30 +45,35 @@ std::mutex cached_casts_mutex;
-
-// autocast_xla_dtype is the lower_precision_fp used by AutocastXLA.
-thread_local at::ScalarType autocast_xla_dtype = at::kBFloat16;
+thread_local std::array<at::ScalarType, at::COMPILE_TIME_MAX_DEVICE_TYPES>
```
is a map easier to maintain than an array? You don't have to match the exact order of dispatch keys, also don't have to specify `undefined` explicitly?

guangyey replied:
I think `map` cannot significantly reduce maintenance efforts, and `array` offers a better performance. Both are OK to me, but I prefer `array` due to the code style.

guangyey committed:
Update on "[WIP] Refactor autocast APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast APIs to be device-agnostic and remove the device-specific autocast APIs.
In C++ side,
`is_enabled()` -> `is_enabled(device_type)`.
`set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
`get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
`set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`
~~`is_cpu_enabled`~~
~~`set_cpu_enabled`~~
~~`get_autocast_cpu_dtype`~~
~~`set_autocast_cpu_dtype`~~
~~`is_xpu_enabled`~~
~~`set_xpu_enabled`~~
~~`get_autocast_xpu_dtype`~~
~~`set_autocast_xpu_dtype`~~
~~`is_ipu_enabled`~~
~~`set_ipu_enabled`~~
~~`get_autocast_ipu_dtype`~~
~~`set_autocast_ipu_dtype`~~
~~`is_hpu_enabled`~~
~~`set_hpu_enabled`~~
~~`get_autocast_hpu_dtype`~~
~~`set_autocast_hpu_dtype`~~
~~`is_xla_enabled`~~
~~`set_xla_enabled`~~
~~`get_autocast_xla_dtype`~~
~~`set_autocast_xla_dtype`~~
~~`is_privateuseone_enabled`~~
~~`set_privateuseone_enabled`~~
~~`get_autocast_privateuseone_dtype`~~
~~`set_autocast_privateuseone_dtype`~~

In Python side,
provide 4 generic autocast APIs:
-`torch.is_autocast_enabled(device_type)`
-`torch.set_autocast_enabled(device_type, new_enabled)`
-`torch.get_autocast_dtype(device_type)`
-`torch.set_autocast_dtype(device_type, dtype)`


cc mcarilli ptrblck leslie-fang-intel jgong5

[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Refactor autocast APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast APIs to be device-agnostic and remove the device-specific autocast APIs.
In C++ side,
`is_enabled()` -> `is_enabled(device_type)`.
`set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
`get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
`set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`
~~`is_cpu_enabled`~~
~~`set_cpu_enabled`~~
~~`get_autocast_cpu_dtype`~~
~~`set_autocast_cpu_dtype`~~
~~`is_xpu_enabled`~~
~~`set_xpu_enabled`~~
~~`get_autocast_xpu_dtype`~~
~~`set_autocast_xpu_dtype`~~
~~`is_ipu_enabled`~~
~~`set_ipu_enabled`~~
~~`get_autocast_ipu_dtype`~~
~~`set_autocast_ipu_dtype`~~
~~`is_hpu_enabled`~~
~~`set_hpu_enabled`~~
~~`get_autocast_hpu_dtype`~~
~~`set_autocast_hpu_dtype`~~
~~`is_xla_enabled`~~
~~`set_xla_enabled`~~
~~`get_autocast_xla_dtype`~~
~~`set_autocast_xla_dtype`~~
~~`is_privateuseone_enabled`~~
~~`set_privateuseone_enabled`~~
~~`get_autocast_privateuseone_dtype`~~
~~`set_autocast_privateuseone_dtype`~~

In Python side,
provide 4 generic autocast APIs:
-`torch.is_autocast_enabled(device_type)`
-`torch.set_autocast_enabled(device_type, new_enabled)`
-`torch.get_autocast_dtype(device_type)`
-`torch.set_autocast_dtype(device_type, dtype)`


cc mcarilli ptrblck leslie-fang-intel jgong5

[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Refactor autocast APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast APIs to be device-agnostic and remove the device-specific autocast APIs.
In C++ side,
`is_enabled()` -> `is_enabled(device_type)`.
`set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
`get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
`set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`
~~`is_cpu_enabled`~~
~~`set_cpu_enabled`~~
~~`get_autocast_cpu_dtype`~~
~~`set_autocast_cpu_dtype`~~
~~`is_xpu_enabled`~~
~~`set_xpu_enabled`~~
~~`get_autocast_xpu_dtype`~~
~~`set_autocast_xpu_dtype`~~
~~`is_ipu_enabled`~~
~~`set_ipu_enabled`~~
~~`get_autocast_ipu_dtype`~~
~~`set_autocast_ipu_dtype`~~
~~`is_hpu_enabled`~~
~~`set_hpu_enabled`~~
~~`get_autocast_hpu_dtype`~~
~~`set_autocast_hpu_dtype`~~
~~`is_xla_enabled`~~
~~`set_xla_enabled`~~
~~`get_autocast_xla_dtype`~~
~~`set_autocast_xla_dtype`~~
~~`is_privateuseone_enabled`~~
~~`set_privateuseone_enabled`~~
~~`get_autocast_privateuseone_dtype`~~
~~`set_autocast_privateuseone_dtype`~~

In Python side,
provide 4 generic autocast APIs:
-`torch.is_autocast_enabled(device_type)`
-`torch.set_autocast_enabled(device_type, new_enabled)`
-`torch.get_autocast_dtype(device_type)`
-`torch.set_autocast_dtype(device_type, dtype)`


cc mcarilli ptrblck leslie-fang-intel jgong5

[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Refactor autocast C++ APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast **C++** APIs to be device-agnostic and deprecate the device-specific autocast  **C++** APIs.
In C++ side,
- `is_enabled()` -> `is_enabled(device_type)`.
- `set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
- `get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
- `set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`

These following C++ APIs are deprecated and should be removed in PyTorch 2.5
- `is_cpu_enabled`
- `set_cpu_enabled`
- `get_autocast_cpu_dtype`
- `set_autocast_cpu_dtype`
- `is_xpu_enabled`
- `set_xpu_enabled`
- `get_autocast_xpu_dtype`
- `set_autocast_xpu_dtype`
- `is_ipu_enabled`
- `set_ipu_enabled`
- `get_autocast_ipu_dtype`
- `set_autocast_ipu_dtype`
- `is_hpu_enabled`
- `set_hpu_enabled`
- `get_autocast_hpu_dtype`
- `set_autocast_hpu_dtype`
- `is_xla_enabled`
- `set_xla_enabled`
- `get_autocast_xla_dtype`
- `set_autocast_xla_dtype`
- `is_privateuseone_enabled`
- `set_privateuseone_enabled`
- `get_autocast_privateuseone_dtype`
- `set_autocast_privateuseone_dtype`

In Python side,
provide 4 generic autocast APIs:
- `torch.is_autocast_enabled(device_type)`
- `torch.set_autocast_enabled(device_type, new_enabled)`
- `torch.get_autocast_dtype(device_type)`
- `torch.set_autocast_dtype(device_type, dtype)`

# Additional Context
We will submit another PR to refactor autocast **Python** APIs based on this PR.

cc mcarilli ptrblck leslie-fang-intel jgong5

[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Refactor autocast C++ APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast **C++** APIs to be device-agnostic and deprecate the device-specific autocast  **C++** APIs.
In C++ side,
- `is_enabled()` -> `is_enabled(device_type)`.
- `set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
- `get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
- `set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`

These following C++ APIs are deprecated and should be removed in PyTorch 2.5
- `is_cpu_enabled`
- `set_cpu_enabled`
- `get_autocast_cpu_dtype`
- `set_autocast_cpu_dtype`
- `is_xpu_enabled`
- `set_xpu_enabled`
- `get_autocast_xpu_dtype`
- `set_autocast_xpu_dtype`
- `is_ipu_enabled`
- `set_ipu_enabled`
- `get_autocast_ipu_dtype`
- `set_autocast_ipu_dtype`
- `is_hpu_enabled`
- `set_hpu_enabled`
- `get_autocast_hpu_dtype`
- `set_autocast_hpu_dtype`
- `is_xla_enabled`
- `set_xla_enabled`
- `get_autocast_xla_dtype`
- `set_autocast_xla_dtype`
- `is_privateuseone_enabled`
- `set_privateuseone_enabled`
- `get_autocast_privateuseone_dtype`
- `set_autocast_privateuseone_dtype`

In Python side,
provide 4 generic autocast APIs:
- `torch.is_autocast_enabled(device_type)`
- `torch.set_autocast_enabled(device_type, new_enabled)`
- `torch.get_autocast_dtype(device_type)`
- `torch.set_autocast_dtype(device_type, dtype)`

# Additional Context
We will submit another PR to refactor autocast **Python** APIs based on this PR.

cc mcarilli ptrblck leslie-fang-intel jgong5

[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Refactor autocast C++ APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast **C++** APIs to be device-agnostic and deprecate the device-specific autocast  **C++** APIs.
In C++ side,
- `is_enabled()` -> `is_enabled(device_type)`.
- `set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
- `get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
- `set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`

These following C++ APIs are deprecated and should be removed in PyTorch 2.5
- `is_cpu_enabled`
- `set_cpu_enabled`
- `get_autocast_cpu_dtype`
- `set_autocast_cpu_dtype`
- `is_xpu_enabled`
- `set_xpu_enabled`
- `get_autocast_xpu_dtype`
- `set_autocast_xpu_dtype`
- `is_ipu_enabled`
- `set_ipu_enabled`
- `get_autocast_ipu_dtype`
- `set_autocast_ipu_dtype`
- `is_hpu_enabled`
- `set_hpu_enabled`
- `get_autocast_hpu_dtype`
- `set_autocast_hpu_dtype`
- `is_xla_enabled`
- `set_xla_enabled`
- `get_autocast_xla_dtype`
- `set_autocast_xla_dtype`
- `is_privateuseone_enabled`
- `set_privateuseone_enabled`
- `get_autocast_privateuseone_dtype`
- `set_autocast_privateuseone_dtype`

In Python side,
provide 4 generic autocast APIs:
- `torch.is_autocast_enabled(device_type)`
- `torch.set_autocast_enabled(device_type, new_enabled)`
- `torch.get_autocast_dtype(device_type)`
- `torch.set_autocast_dtype(device_type, dtype)`

# Additional Context
We will submit another PR to refactor autocast **Python** APIs based on this PR.

cc mcarilli ptrblck leslie-fang-intel jgong5

[ghstack-poisoned]

albanD commented on aten/src/ATen/autocast_mode.cpp:
```diff
@@ -91,30 +45,35 @@ std::mutex cached_casts_mutex;
-// autocast_xla_dtype is the lower_precision_fp used by AutocastXLA.
-thread_local at::ScalarType autocast_xla_dtype = at::kBFloat16;
+thread_local std::array<at::ScalarType, at::COMPILE_TIME_MAX_DEVICE_TYPES>
+    autocast_dtype = {
```
I guess this is an exact match of https://github.com/pytorch/pytorch/blob/25c65d664254ca48cb909375d90652dab60f8d44/c10/core/DeviceType.h#L36-L63 in terms of how to index into it?

Can you add a comment to that effect and a static assert that they're the same size?

guangyey replied:
Add comments and static_assert to ensure their match between autocast data type and device type.

albanD reviewed (commented):
Really cool!

albanD reviewed (approved):
Putting an accept conditional on the new comment and compile-time assert.

guangyey commented:
> Putting an accept conditional on the new comment and compile-time assert.

Add some deprecated messages both on C++ and Python APIs at run time.

guangyey committed:
Update on "Refactor autocast C++ APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast **C++** APIs to be device-agnostic and deprecate the device-specific autocast  **C++** APIs.
In C++ side,
- `is_enabled()` -> `is_enabled(device_type)`.
- `set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
- `get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
- `set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`

These following C++ APIs are deprecated and should be removed in PyTorch 2.5
- `is_cpu_enabled`
- `set_cpu_enabled`
- `get_autocast_cpu_dtype`
- `set_autocast_cpu_dtype`
- `is_xpu_enabled`
- `set_xpu_enabled`
- `get_autocast_xpu_dtype`
- `set_autocast_xpu_dtype`
- `is_ipu_enabled`
- `set_ipu_enabled`
- `get_autocast_ipu_dtype`
- `set_autocast_ipu_dtype`
- `is_hpu_enabled`
- `set_hpu_enabled`
- `get_autocast_hpu_dtype`
- `set_autocast_hpu_dtype`
- `is_xla_enabled`
- `set_xla_enabled`
- `get_autocast_xla_dtype`
- `set_autocast_xla_dtype`
- `is_privateuseone_enabled`
- `set_privateuseone_enabled`
- `get_autocast_privateuseone_dtype`
- `set_autocast_privateuseone_dtype`

In Python side,
provide 4 generic autocast APIs:
- `torch.is_autocast_enabled(device_type)`
- `torch.set_autocast_enabled(device_type, new_enabled)`
- `torch.get_autocast_dtype(device_type)`
- `torch.set_autocast_dtype(device_type, dtype)`

# Additional Context
We will submit another PR to refactor autocast **Python** APIs based on this PR.

cc mcarilli ptrblck leslie-fang-intel jgong5

[ghstack-poisoned]

guangyey committed:
Update on "Refactor autocast C++ APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast **C++** APIs to be device-agnostic and deprecate the device-specific autocast  **C++** APIs.
In C++ side,
- `is_enabled()` -> `is_enabled(device_type)`.
- `set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
- `get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
- `set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`

These following C++ APIs are deprecated and should be removed in PyTorch 2.5
- `is_cpu_enabled`
- `set_cpu_enabled`
- `get_autocast_cpu_dtype`
- `set_autocast_cpu_dtype`
- `is_xpu_enabled`
- `set_xpu_enabled`
- `get_autocast_xpu_dtype`
- `set_autocast_xpu_dtype`
- `is_ipu_enabled`
- `set_ipu_enabled`
- `get_autocast_ipu_dtype`
- `set_autocast_ipu_dtype`
- `is_hpu_enabled`
- `set_hpu_enabled`
- `get_autocast_hpu_dtype`
- `set_autocast_hpu_dtype`
- `is_xla_enabled`
- `set_xla_enabled`
- `get_autocast_xla_dtype`
- `set_autocast_xla_dtype`
- `is_privateuseone_enabled`
- `set_privateuseone_enabled`
- `get_autocast_privateuseone_dtype`
- `set_autocast_privateuseone_dtype`

In Python side,
provide 4 generic autocast APIs:
- `torch.is_autocast_enabled(device_type)`
- `torch.set_autocast_enabled(device_type, new_enabled)`
- `torch.get_autocast_dtype(device_type)`
- `torch.set_autocast_dtype(device_type, dtype)`

# Additional Context
We will submit another PR to refactor autocast **Python** APIs based on this PR.

cc mcarilli ptrblck leslie-fang-intel jgong5

[ghstack-poisoned]

guangyey committed:
Update on "Refactor autocast C++ APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast **C++** APIs to be device-agnostic and deprecate the device-specific autocast  **C++** APIs.
In C++ side,
- `is_enabled()` -> `is_enabled(device_type)`.
- `set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
- `get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
- `set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`

These following C++ APIs are deprecated and should be removed in PyTorch 2.5
- `is_cpu_enabled`
- `set_cpu_enabled`
- `get_autocast_cpu_dtype`
- `set_autocast_cpu_dtype`
- `is_xpu_enabled`
- `set_xpu_enabled`
- `get_autocast_xpu_dtype`
- `set_autocast_xpu_dtype`
- `is_ipu_enabled`
- `set_ipu_enabled`
- `get_autocast_ipu_dtype`
- `set_autocast_ipu_dtype`
- `is_hpu_enabled`
- `set_hpu_enabled`
- `get_autocast_hpu_dtype`
- `set_autocast_hpu_dtype`
- `is_xla_enabled`
- `set_xla_enabled`
- `get_autocast_xla_dtype`
- `set_autocast_xla_dtype`
- `is_privateuseone_enabled`
- `set_privateuseone_enabled`
- `get_autocast_privateuseone_dtype`
- `set_autocast_privateuseone_dtype`

In Python side,
provide 4 generic autocast APIs:
- `torch.is_autocast_enabled(device_type)`
- `torch.set_autocast_enabled(device_type, new_enabled)`
- `torch.get_autocast_dtype(device_type)`
- `torch.set_autocast_dtype(device_type, dtype)`

# Additional Context
We will submit another PR to refactor autocast **Python** APIs based on this PR.

cc mcarilli ptrblck leslie-fang-intel jgong5

[ghstack-poisoned]

guangyey commented on torch/_dynamo/output_graph.py:
```diff
@@ -595,21 +595,30 @@ def save_global_state(self, out=None):
         global_state["autocast_enabled"] = (
-            torch.set_autocast_enabled,
-            torch.is_autocast_enabled(),
+            autocast_specific_backend("cuda", torch.set_autocast_enabled),
```
Some dynamo-related UTs can catch warnings as errors. So it results in UT failures if we use deprecated APIs here.

guangyey commented on torch/utils/checkpoint.py:
```diff
@@ -194,25 +194,16 @@ def set_device_states(devices, states) -> None:
-        device_autocast_kwargs = None
+    assert device != "cpu"
+    device_autocast_kwargs = {
+        "enabled": torch.is_autocast_enabled(device),
```
The deprecated APIs will raise a warning as an error in some dynamo-related UTs. So use device-agnostic APIs instead.

guangyey committed:
Update on "Refactor autocast C++ APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast **C++** APIs to be device-agnostic and deprecate the device-specific autocast  **C++** APIs.
In C++ side,
- `is_enabled()` -> `is_enabled(device_type)`.
- `set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
- `get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
- `set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`

These following C++ APIs are deprecated and should be removed in PyTorch 2.5
- `is_cpu_enabled`
- `set_cpu_enabled`
- `get_autocast_cpu_dtype`
- `set_autocast_cpu_dtype`
- `is_xpu_enabled`
- `set_xpu_enabled`
- `get_autocast_xpu_dtype`
- `set_autocast_xpu_dtype`
- `is_ipu_enabled`
- `set_ipu_enabled`
- `get_autocast_ipu_dtype`
- `set_autocast_ipu_dtype`
- `is_hpu_enabled`
- `set_hpu_enabled`
- `get_autocast_hpu_dtype`
- `set_autocast_hpu_dtype`
- `is_xla_enabled`
- `set_xla_enabled`
- `get_autocast_xla_dtype`
- `set_autocast_xla_dtype`
- `is_privateuseone_enabled`
- `set_privateuseone_enabled`
- `get_autocast_privateuseone_dtype`
- `set_autocast_privateuseone_dtype`

In Python side,
provide 4 generic autocast APIs:
- `torch.is_autocast_enabled(device_type)`
- `torch.set_autocast_enabled(device_type, new_enabled)`
- `torch.get_autocast_dtype(device_type)`
- `torch.set_autocast_dtype(device_type, dtype)`

# Additional Context
We will submit another PR to refactor autocast **Python** APIs based on this PR.

cc mcarilli ptrblck leslie-fang-intel jgong5 voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

guangyey committed:
Update on "Refactor autocast C++ APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast **C++** APIs to be device-agnostic and deprecate the device-specific autocast  **C++** APIs.
In C++ side,
- `is_enabled()` -> `is_enabled(device_type)`.
- `set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
- `get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
- `set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`

These following C++ APIs are deprecated and should be removed in PyTorch 2.5
- `is_cpu_enabled`
- `set_cpu_enabled`
- `get_autocast_cpu_dtype`
- `set_autocast_cpu_dtype`
- `is_xpu_enabled`
- `set_xpu_enabled`
- `get_autocast_xpu_dtype`
- `set_autocast_xpu_dtype`
- `is_ipu_enabled`
- `set_ipu_enabled`
- `get_autocast_ipu_dtype`
- `set_autocast_ipu_dtype`
- `is_hpu_enabled`
- `set_hpu_enabled`
- `get_autocast_hpu_dtype`
- `set_autocast_hpu_dtype`
- `is_xla_enabled`
- `set_xla_enabled`
- `get_autocast_xla_dtype`
- `set_autocast_xla_dtype`
- `is_privateuseone_enabled`
- `set_privateuseone_enabled`
- `get_autocast_privateuseone_dtype`
- `set_autocast_privateuseone_dtype`

In Python side,
provide 4 generic autocast APIs:
- `torch.is_autocast_enabled(device_type)`
- `torch.set_autocast_enabled(device_type, new_enabled)`
- `torch.get_autocast_dtype(device_type)`
- `torch.set_autocast_dtype(device_type, dtype)`

# Additional Context
We will submit another PR to refactor autocast **Python** APIs based on this PR.

cc mcarilli ptrblck leslie-fang-intel jgong5 voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

guangyey committed:
Update on "Refactor autocast C++ APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast **C++** APIs to be device-agnostic and deprecate the device-specific autocast  **C++** APIs.
In C++ side,
- `is_enabled()` -> `is_enabled(device_type)`.
- `set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
- `get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
- `set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`

These following C++ APIs are deprecated and should be removed in PyTorch 2.5
- `is_cpu_enabled`
- `set_cpu_enabled`
- `get_autocast_cpu_dtype`
- `set_autocast_cpu_dtype`
- `is_xpu_enabled`
- `set_xpu_enabled`
- `get_autocast_xpu_dtype`
- `set_autocast_xpu_dtype`
- `is_ipu_enabled`
- `set_ipu_enabled`
- `get_autocast_ipu_dtype`
- `set_autocast_ipu_dtype`
- `is_hpu_enabled`
- `set_hpu_enabled`
- `get_autocast_hpu_dtype`
- `set_autocast_hpu_dtype`
- `is_xla_enabled`
- `set_xla_enabled`
- `get_autocast_xla_dtype`
- `set_autocast_xla_dtype`
- `is_privateuseone_enabled`
- `set_privateuseone_enabled`
- `get_autocast_privateuseone_dtype`
- `set_autocast_privateuseone_dtype`

In Python side,
provide 4 generic autocast APIs:
- `torch.is_autocast_enabled(device_type)`
- `torch.set_autocast_enabled(device_type, new_enabled)`
- `torch.get_autocast_dtype(device_type)`
- `torch.set_autocast_dtype(device_type, dtype)`

# Additional Context
We will submit another PR to refactor autocast **Python** APIs based on this PR.

cc mcarilli ptrblck leslie-fang-intel jgong5 voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

guangyey committed:
Update on "Refactor autocast C++ APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast **C++** APIs to be device-agnostic and deprecate the device-specific autocast  **C++** APIs.
In C++ side,
- `is_enabled()` -> `is_enabled(device_type)`.
- `set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
- `get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
- `set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`

These following C++ APIs are deprecated and should be removed in PyTorch 2.5
- `is_cpu_enabled`
- `set_cpu_enabled`
- `get_autocast_cpu_dtype`
- `set_autocast_cpu_dtype`
- `is_xpu_enabled`
- `set_xpu_enabled`
- `get_autocast_xpu_dtype`
- `set_autocast_xpu_dtype`
- `is_ipu_enabled`
- `set_ipu_enabled`
- `get_autocast_ipu_dtype`
- `set_autocast_ipu_dtype`
- `is_hpu_enabled`
- `set_hpu_enabled`
- `get_autocast_hpu_dtype`
- `set_autocast_hpu_dtype`
- `is_xla_enabled`
- `set_xla_enabled`
- `get_autocast_xla_dtype`
- `set_autocast_xla_dtype`
- `is_privateuseone_enabled`
- `set_privateuseone_enabled`
- `get_autocast_privateuseone_dtype`
- `set_autocast_privateuseone_dtype`

In Python side,
provide 4 generic autocast APIs:
- `torch.is_autocast_enabled(device_type)`
- `torch.set_autocast_enabled(device_type, new_enabled)`
- `torch.get_autocast_dtype(device_type)`
- `torch.set_autocast_dtype(device_type, dtype)`

# Additional Context
We will submit another PR to refactor autocast **Python** APIs based on this PR.

cc mcarilli ptrblck leslie-fang-intel jgong5 voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

guangyey committed:
Update on "Refactor autocast C++ APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast **C++** APIs to be device-agnostic and deprecate the device-specific autocast  **C++** APIs.
In C++ side,
- `is_enabled()` -> `is_enabled(device_type)`.
- `set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
- `get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
- `set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`

These following C++ APIs are deprecated and should be removed in PyTorch 2.5
- `is_cpu_enabled`
- `set_cpu_enabled`
- `get_autocast_cpu_dtype`
- `set_autocast_cpu_dtype`
- `is_xpu_enabled`
- `set_xpu_enabled`
- `get_autocast_xpu_dtype`
- `set_autocast_xpu_dtype`
- `is_ipu_enabled`
- `set_ipu_enabled`
- `get_autocast_ipu_dtype`
- `set_autocast_ipu_dtype`
- `is_hpu_enabled`
- `set_hpu_enabled`
- `get_autocast_hpu_dtype`
- `set_autocast_hpu_dtype`
- `is_xla_enabled`
- `set_xla_enabled`
- `get_autocast_xla_dtype`
- `set_autocast_xla_dtype`
- `is_privateuseone_enabled`
- `set_privateuseone_enabled`
- `get_autocast_privateuseone_dtype`
- `set_autocast_privateuseone_dtype`

In Python side,
provide 4 generic autocast APIs:
- `torch.is_autocast_enabled(device_type)`
- `torch.set_autocast_enabled(device_type, new_enabled)`
- `torch.get_autocast_dtype(device_type)`
- `torch.set_autocast_dtype(device_type, dtype)`

# Additional Context
We will submit another PR to refactor autocast **Python** APIs based on this PR.

cc mcarilli ptrblck leslie-fang-intel jgong5 voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

guangyey committed:
Update on "Refactor autocast C++ APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast **C++** APIs to be device-agnostic and deprecate the device-specific autocast  **C++** APIs.
In C++ side,
- `is_enabled()` -> `is_enabled(device_type)`.
- `set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
- `get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
- `set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`

These following C++ APIs are deprecated and should be removed in PyTorch 2.5
- `is_cpu_enabled`
- `set_cpu_enabled`
- `get_autocast_cpu_dtype`
- `set_autocast_cpu_dtype`
- `is_xpu_enabled`
- `set_xpu_enabled`
- `get_autocast_xpu_dtype`
- `set_autocast_xpu_dtype`
- `is_ipu_enabled`
- `set_ipu_enabled`
- `get_autocast_ipu_dtype`
- `set_autocast_ipu_dtype`
- `is_hpu_enabled`
- `set_hpu_enabled`
- `get_autocast_hpu_dtype`
- `set_autocast_hpu_dtype`
- `is_xla_enabled`
- `set_xla_enabled`
- `get_autocast_xla_dtype`
- `set_autocast_xla_dtype`
- `is_privateuseone_enabled`
- `set_privateuseone_enabled`
- `get_autocast_privateuseone_dtype`
- `set_autocast_privateuseone_dtype`

In Python side,
provide 4 generic autocast APIs:
- `torch.is_autocast_enabled(device_type)`
- `torch.set_autocast_enabled(device_type, new_enabled)`
- `torch.get_autocast_dtype(device_type)`
- `torch.set_autocast_dtype(device_type, dtype)`

# Additional Context
We will submit another PR to refactor autocast **Python** APIs based on this PR.

cc mcarilli ptrblck leslie-fang-intel jgong5 voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

guangyey committed:
Update on "Refactor autocast C++ APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast **C++** APIs to be device-agnostic and deprecate the device-specific autocast  **C++** APIs.
In C++ side,
- `is_enabled()` -> `is_enabled(device_type)`.
- `set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
- `get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
- `set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`

These following C++ APIs are deprecated and should be removed in PyTorch 2.5
- `is_cpu_enabled`
- `set_cpu_enabled`
- `get_autocast_cpu_dtype`
- `set_autocast_cpu_dtype`
- `is_xpu_enabled`
- `set_xpu_enabled`
- `get_autocast_xpu_dtype`
- `set_autocast_xpu_dtype`
- `is_ipu_enabled`
- `set_ipu_enabled`
- `get_autocast_ipu_dtype`
- `set_autocast_ipu_dtype`
- `is_hpu_enabled`
- `set_hpu_enabled`
- `get_autocast_hpu_dtype`
- `set_autocast_hpu_dtype`
- `is_xla_enabled`
- `set_xla_enabled`
- `get_autocast_xla_dtype`
- `set_autocast_xla_dtype`
- `is_privateuseone_enabled`
- `set_privateuseone_enabled`
- `get_autocast_privateuseone_dtype`
- `set_autocast_privateuseone_dtype`

In Python side,
provide 4 generic autocast APIs:
- `torch.is_autocast_enabled(device_type)`
- `torch.set_autocast_enabled(device_type, new_enabled)`
- `torch.get_autocast_dtype(device_type)`
- `torch.set_autocast_dtype(device_type, dtype)`

# Additional Context
We will submit another PR to refactor autocast **Python** APIs based on this PR.

cc mcarilli ptrblck leslie-fang-intel jgong5 voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

guangyey committed:
Update on "Refactor autocast C++ APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast **C++** APIs to be device-agnostic and deprecate the device-specific autocast  **C++** APIs.
In C++ side,
- `is_enabled()` -> `is_enabled(device_type)`.
- `set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
- `get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
- `set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`

These following C++ APIs are deprecated and should be removed in PyTorch 2.5
- `is_cpu_enabled`
- `set_cpu_enabled`
- `get_autocast_cpu_dtype`
- `set_autocast_cpu_dtype`
- `is_xpu_enabled`
- `set_xpu_enabled`
- `get_autocast_xpu_dtype`
- `set_autocast_xpu_dtype`
- `is_ipu_enabled`
- `set_ipu_enabled`
- `get_autocast_ipu_dtype`
- `set_autocast_ipu_dtype`
- `is_hpu_enabled`
- `set_hpu_enabled`
- `get_autocast_hpu_dtype`
- `set_autocast_hpu_dtype`
- `is_xla_enabled`
- `set_xla_enabled`
- `get_autocast_xla_dtype`
- `set_autocast_xla_dtype`
- `is_privateuseone_enabled`
- `set_privateuseone_enabled`
- `get_autocast_privateuseone_dtype`
- `set_autocast_privateuseone_dtype`

In Python side,
provide 4 generic autocast APIs:
- `torch.is_autocast_enabled(device_type)`
- `torch.set_autocast_enabled(device_type, new_enabled)`
- `torch.get_autocast_dtype(device_type)`
- `torch.set_autocast_dtype(device_type, dtype)`

# Additional Context
We will submit another PR to refactor autocast **Python** APIs based on this PR.

cc mcarilli ptrblck leslie-fang-intel jgong5 voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

guangyey committed:
Update on "Refactor autocast C++ APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast **C++** APIs to be device-agnostic and deprecate the device-specific autocast  **C++** APIs.
In C++ side,
- `is_enabled()` -> `is_enabled(device_type)`.
- `set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
- `get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
- `set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`

These following C++ APIs are deprecated and should be removed in PyTorch 2.5
- `is_cpu_enabled`
- `set_cpu_enabled`
- `get_autocast_cpu_dtype`
- `set_autocast_cpu_dtype`
- `is_xpu_enabled`
- `set_xpu_enabled`
- `get_autocast_xpu_dtype`
- `set_autocast_xpu_dtype`
- `is_ipu_enabled`
- `set_ipu_enabled`
- `get_autocast_ipu_dtype`
- `set_autocast_ipu_dtype`
- `is_hpu_enabled`
- `set_hpu_enabled`
- `get_autocast_hpu_dtype`
- `set_autocast_hpu_dtype`
- `is_xla_enabled`
- `set_xla_enabled`
- `get_autocast_xla_dtype`
- `set_autocast_xla_dtype`
- `is_privateuseone_enabled`
- `set_privateuseone_enabled`
- `get_autocast_privateuseone_dtype`
- `set_autocast_privateuseone_dtype`

In Python side,
provide 4 generic autocast APIs:
- `torch.is_autocast_enabled(device_type)`
- `torch.set_autocast_enabled(device_type, new_enabled)`
- `torch.get_autocast_dtype(device_type)`
- `torch.set_autocast_dtype(device_type, dtype)`

# Additional Context
We will submit another PR to refactor autocast **Python** APIs based on this PR.

cc mcarilli ptrblck leslie-fang-intel jgong5 voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

guangyey commented on torch/_functorch/_aot_autograd/utils.py:
```diff
@@ -79,10 +79,10 @@ def normalize_as_list(x):
-        torch.is_autocast_cpu_enabled(),
-        torch.get_autocast_gpu_dtype(),
-        torch.get_autocast_cpu_dtype(),
+        torch.is_autocast_enabled("cuda"),
```
fix UT warnings check issue on `xla/test/test_dynamo.py::test_all_cpu_tensor`

guangyey commented:
@pytorchbot merge

guangyey committed:
Update on "Refactor autocast C++ APIs to be device-agnostic"


# Motivation
This PR aims to refactor autocast **C++** APIs to be device-agnostic and deprecate the device-specific autocast  **C++** APIs.
In C++ side,
- `is_enabled()` -> `is_enabled(device_type)`.
- `set_enabled(new_enabled)` -> `set_enabled(device_type, new_enabled)`.
- `get_autocast_dtype()` -> `get_autocast_dtype(device_type)`
- `set_autocast_dtype(dtype)` -> `set_autocast_dtype(device_type, dtype)`

These following C++ APIs are deprecated and should be removed in PyTorch 2.5
- `is_cpu_enabled`
- `set_cpu_enabled`
- `get_autocast_cpu_dtype`
- `set_autocast_cpu_dtype`
- `is_xpu_enabled`
- `set_xpu_enabled`
- `get_autocast_xpu_dtype`
- `set_autocast_xpu_dtype`
- `is_ipu_enabled`
- `set_ipu_enabled`
- `get_autocast_ipu_dtype`
- `set_autocast_ipu_dtype`
- `is_hpu_enabled`
- `set_hpu_enabled`
- `get_autocast_hpu_dtype`
- `set_autocast_hpu_dtype`
- `is_xla_enabled`
- `set_xla_enabled`
- `get_autocast_xla_dtype`
- `set_autocast_xla_dtype`
- `is_privateuseone_enabled`
- `set_privateuseone_enabled`
- `get_autocast_privateuseone_dtype`
- `set_autocast_privateuseone_dtype`

In Python side,
provide 4 generic autocast APIs:
- `torch.is_autocast_enabled(device_type)`
- `torch.set_autocast_enabled(device_type, new_enabled)`
- `torch.get_autocast_dtype(device_type)`
- `torch.set_autocast_dtype(device_type, dtype)`

# Additional Context
We will submit another PR to refactor autocast **Python** APIs based on this PR.

cc mcarilli ptrblck leslie-fang-intel jgong5 voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]
